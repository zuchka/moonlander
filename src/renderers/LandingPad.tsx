import React from 'react';
import { StyleSheet } from 'react-native';
import Matter from 'matter-js';
import Svg, { Rect } from 'react-native-svg';

interface LandingPadProps {
    body: Matter.Body;
    size: number[];
    testID?: string; // For the Svg container
    rectTestID?: string; // For the Rect
}

const LandingPad: React.FC<LandingPadProps> = ({ body, size, testID, rectTestID }) => {
    const width = size[0];
    const height = size[1];
    const { x: bodyX, y: bodyY } = body.position;

    // Calculate top-left position
    const topLeftX = bodyX - width / 2;
    const topLeftY = bodyY - height / 2;

    return (
        <Svg
            testID={testID}
            width={width}
            height={height}
            style={[
                styles.landingPadSvg,
                {
                    left: topLeftX,
                    top: topLeftY,
                    width: width,
                    height: height,
                }
            ]}
        >
            <Rect
                testID={rectTestID}
                x="0" // Positioned relative to the Svg container's top-left
                y="0"
                width={width}
                height={height}
                fill="mediumseagreen" // Example color
            />
        </Svg>
    );
};

const styles = StyleSheet.create({
    landingPadSvg: {
        position: 'absolute',
        // backgroundColor: 'rgba(0, 0, 255, 0.1)', // Optional: debug bounds
    },
});

export default LandingPad; 