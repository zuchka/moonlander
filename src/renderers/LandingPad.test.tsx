import React from 'react';
import { render } from '@testing-library/react-native';
import LandingPad from './LandingPad'; // Revert path back to relative
import Matter from 'matter-js';
import Svg, { Rect } from 'react-native-svg';

// Mock Body
const mockBody: Matter.Body = {
    position: { x: 400, y: 550 },
    angle: 0,
} as Matter.Body;

const size = [80, 10]; // width, height

describe('<LandingPad /> Renderer', () => {

    test('renders an Svg container with correct position and size', () => {
        const { getByTestId } = render(<LandingPad body={mockBody} size={size} testID="pad-svg" />);
        const svg = getByTestId('pad-svg');

        expect(svg).toBeDefined();
        // Remove fragile style check - rely on snapshot and existence check
        // expect(svg.props.style).toEqual(expect.arrayContaining([
        //     expect.objectContaining({
        //         position: 'absolute',
        //         left: 360,
        //         top: 545,
        //         width: size[0],
        //         height: size[1],
        //     })
        // ]));
        // Basic check that width/height props are passed
        expect(svg.props.width).toBe(size[0]);
        expect(svg.props.height).toBe(size[1]);
    });

    test('renders a Rect element inside', () => {
        const { getByTestId } = render(
            <LandingPad body={mockBody} size={size} testID="pad-svg" rectTestID="pad-rect" />
        );
        const rect = getByTestId('pad-rect');

        expect(rect).toBeDefined();
        expect(rect.props.width).toBe(size[0]);
        expect(rect.props.height).toBe(size[1]);
        expect(rect.props.fill).toBeDefined(); // Check styling
    });

    // Optional: Snapshot test
    test('matches snapshot', () => {
        const tree = render(<LandingPad body={mockBody} size={size} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

// Helper type
type LandingPadProps = {
  body: Matter.Body;
  size: number[];
  testID?: string;
  rectTestID?: string;
}; 