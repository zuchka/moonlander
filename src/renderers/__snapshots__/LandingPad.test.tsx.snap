// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LandingPad /> Renderer matches snapshot 1`] = `
<RNSVGSvgView
  bbHeight={10}
  bbWidth={80}
  focusable={false}
  height={10}
  style={
    [
      {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      [
        {
          "position": "absolute",
        },
        {
          "height": 10,
          "left": 360,
          "top": 545,
          "width": 80,
        },
      ],
      {
        "flex": 0,
        "height": 10,
        "width": 80,
      },
    ]
  }
  width={80}
>
  <RNSVGGroup
    fill={
      {
        "payload": 4278190080,
        "type": 0,
      }
    }
  >
    <RNSVGRect
      fill={
        {
          "payload": 4282168177,
          "type": 0,
        }
      }
      height={10}
      propList={
        [
          "fill",
        ]
      }
      width={80}
      x="0"
      y="0"
    />
  </RNSVGGroup>
</RNSVGSvgView>
`;
